tiledb/sm/array_schema/dimension.cc

```diff
@@ -1331,6 +1395,45 @@ ByteVecValue Dimension::map_from_uint64(
   return ret;
 }
 
+template <>
+ByteVecValue Dimension::map_from_uint64<double>(
+    const Dimension* dim, uint64_t value, int bits, uint64_t max_bucket_val) {
+  assert(dim != nullptr);
+  assert(!dim->domain().empty());
+  (void)bits;  // Not needed here
+
+  ByteVecValue ret(sizeof(double));
+
+  auto dom_start = *(const double*)dim->domain().start() / 2;
+  auto dom_end = *(const double*)dim->domain().end() / 2;
+  double dom_range = dom_end - dom_start;
+
+  // Essentially take the largest value in the bucket
+  double norm_coord = ((value + 1) / (double)max_bucket_val) * dom_range;
+  double coord = norm_coord + dom_start;
+
+  coord *= 2;
+
+  /**
+   * Here we should have the minimum value for the next bucket. But because of
+   * rounding errors, std::nextafter might not return what is going to be
+   * the maximum value for this bucket, a few calls might be necessary to
+   * find the correct value. This will recalculate the bucket as in the
+   * map_to_uint64 functions and make sure we find the exact value we are
+   * after.
+   */
+  uint64_t bucket;
+  do {
```

Comment:
Is this guaranteed to converge? If so, (1) do we have a proof of correctness, and (2) do we know the worst case number of iterations?

Reply:
In what I have seen, this does converge fast but good point we should enforce a maximum of iterations to prevent hangs. Better to fail the partitioning later than create a hang. We could use a BigDecimal library here but since this is not called very often this might be ok.